complex/src/scalar.rs

```diff
@@ -0,0 +1,294 @@
+
+use traits::{Num, FromPrimitive, Float};
+use std::ops::Neg;
+
+use Complex;
+
+pub trait Scalar: Num + Copy + Neg<Output = Self> {
+    /// Associated Real type
+    type Real: Scalar<Real = Self::Real>;
```

Comment:
I like the self-referential constraint!  It seems like the kind of thing the compiler might not be able to untangle, but it looks like that worked out.

If we add the associated `type Complex` back, can it be nailed down to the concrete `Complex` type?  Something like:

```rust
type Real: Scalar<Real = Self::Real, Complex = Complex<Self::Real>>;
type Complex: Scalar<Real = Self::Real, Complex = Complex<Self::Real>>;
```

Reply:
When I think more about it, I'm torn whether we should allow other `Complex` types or not.  In theory, someone could want to implement their own type, maybe even just a newtype wrapper.  Then these self-referential constraints would be a burden, e.g. `MyComplex<f64>` would also have to wrap the `Real` type to meet the constraits, since with `Real = f64`, `f64::Complex` would not be `MyComplex`.

What do you think?